IK-1482 (#1)
imagekit-developer/imagekit-next: README.md

[<img width="250" alt="ImageKit.io" src="https://raw.githubusercontent.com/imagekit-developer/imagekit-javascript/master/assets/imagekit-light-logo.svg"/>](https://imagekit.io)

# ImageKit.io Next.js SDK

[![Node CI](https://github.com/imagekit-developer/imagekit-next/workflows/Node%20CI/badge.svg)](https://github.com/imagekit-developer/imagekit-next/)
[![npm version](https://img.shields.io/npm/v/imagekit-next)](https://www.npmjs.com/package/imagekit-next)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Twitter Follow](https://img.shields.io/twitter/follow/imagekitio?label=Follow&style=social)](https://twitter.com/ImagekitIo)

ImageKit Next.js SDK allows you to resize, optimize, deliver, and upload images and videos in your Next.js application.

> ImageKit is a complete media storage, optimization, and transformation solution that comes with an image and video CDN. It can be integrated with your existing infrastructure - storage like AWS S3, web servers, your CDN, and custom domain names, allowing you to deliver optimized images in minutes with minimal code changes.

Author: ```suggestion
[ImageKit](https://imagekit.io) is a complete media storage, optimization, and transformation solution that comes with an image and video CDN. It can be integrated with your existing infrastructure - storage like AWS S3, web servers, your CDN, and custom domain names, allowing you to deliver optimized images in minutes with minimal code changes.
```

Reviewer: done

> ## Installation
> 

Author: Add an introductory line here like 

```suggestion
## Installation

Add `imagekit-next` to your project by executing one of the following commands:

```

Reviewer: done

> 
> ```shell
> npm install --save imagekit-next
> ```
> 
> or
> 
> ```shell
> yarn add imagekit-next
> ```

Author: ```suggestion

All you need to do to start is to install the `imagekit-next` package

```shell
npm install --save imagekit-next
```

or

```shell
yarn add imagekit-next
```
```

## Usage

### Initialization

Import components in your code:

```js
import { IKImage, IKVideo, ImageKitProvider, IKUpload, ImageKitContext } from 'imagekit-next'
```

### Pages Router (/src)

By default, `imagekit-next` fully supports the included components in the `src` directory without any additional setup.

### App Router (/app)

To use `imagekit-next` in the app directory, you currently need to mark the parent page or component as a Client Component.

This requirement exists because components are divided into two types: Client Components and Server Components.

To do this, simply add the following at the top of the file:

```js
"use client";
```

> 

Author: ### App Router (/app)

The components in the app directory are by default React Server Side Components (RSCs), but not everything can be done server-side. That's where Client components come in.

`imagekit-next` provides components that operate exclusively on client side.

When using the sdk for the first time, you might enter either of the following errors
```
Error: (0 , react__WEBPACK_IMPORTED_MODULE_0__.createContext) is not a function
```
or 
```
You're importing a component that needs createContext. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.
```

To simply fix this, you can force the component or the page to operate solely on Client Side by adding `"use client";` directive at the start of the file, you can read more about it here [Using Client Components in Next.js](https://nextjs.org/docs/app/building-your-application/rendering/client-components#using-client-components-in-nextjs)

Reviewer: done

> ### Configure hostname
> 
> To protect your application from malicious users, configuration is required in order to use external images. This ensures that only external images from your account can be served from the Next.js Image Optimization API.

Author: ```suggestion
### Allowing images to be loaded from trusted external URLs

To protect your application from malicious users, configuration is required in order to use external images. The `remotePatterns` configuration in Next.js specifies the allowed sources for images loaded via external URLs. This enhances security by preventing unauthorized image sources, which could potentially expose the application to vulnerabilities like cross-site scripting (XSS) or data theft. By defining a whitelist of trusted image sources, developers can ensure that only images from these sources are loaded, thereby safeguarding the application. For more details, refer to the [Next.js documentation](https://nextjs.org/docs/app/api-reference/components/image#remotepatterns).```

Reviewer: done

`next.config.js`
```js
module.exports = {
  images: {
    remotePatterns: [
      {
        protocol: "https",
        hostname: "ik.imagekit.io",
        port: '',
      },
    ],
  },
}
```

> ### Quick examples

Author: ```suggestion
### Examples
```

Reviewer: done

```js
// Ensure you pass publicKey, urlEndpoint, and authenticator function to the parent ImageKitProvider component or to the IKUpload component directly.
<ImageKitProvider
  publicKey="your_public_api_key"
  urlEndpoint="https://ik.imagekit.io/your_imagekit_id"
  authenticator={()=>Promise} 
  // This promise  resolves with an object containing the necessary security parameters i.e `signature`, `token`, and `expire`.
  >
  <IKUpload
    onError={onError}
    onSuccess={onSuccess}
  />

  // Passing different upload API options
  <IKUpload
    fileName="file-name.jpg"
    tags={["sample-tag1", "sample-tag2"]}
    customCoordinates={"10,10,10,10"}
    isPrivateFile={false}
    useUniqueFileName={true}
    responseFields={["tags"]}
    folder={"/sample-folder"}
    ref={uploadRef}
    onError={onError} onSuccess={onSuccess}
    transformation={{
      pre: 'l-text,i-Imagekit,fs-50,l-end', 
      post: [
        {
          'type': 'transformation', 
          'value': 'w-100'
        }
      ]
    }}
  />
</ImageKitProvider>
```

## Demo application
* The official step-by-step Next.js quick start guide - https://docs.imagekit.io/getting-started/quickstart-guides/next

> * The official step-by-step Next.js quick start guide - https://docs.imagekit.io/getting-started/quickstart-guides/next

Author: this is duplicate

Reviewer: done removed duplicate one

     Low-quality image placeholder
     Will first load https://ik.imagekit.io/your_imagekit_id/tr:h-200,w-200:q-20,bl-6/default-image.jpg, while the original image, i.e., https://ik.imagekit.io/your_imagekit_id/tr:h-200,w-200/default-image.jpg is being loaded in the background.
  */
  <IKImage
    path="/default-image.jpg"
    transformation={[{
      "height": "200",
      "width": "200"
    }]}
    lqip={{ active: true }}
    alt="Alt text"
  />

  // Low-quality image placeholder with custom quality and blur values
  <IKImage
    path="/default-image.jpg"
    transformation={[{
      "height": "200",
      "width": "200"
    }]}
    lqip={{ active: true, quality: 20, blur: 10 }}
    alt="Alt text"
  />

  // Low-quality image placeholder and lazy loading of original image in the background
  <IKImage
    path="/default-image.jpg"
    transformation={[{
      "height": "200",
      "width": "200"
    }]}
    loading="lazy"
    lqip={{ active: true }}
    alt="Alt text"
  />
</ImageKitProvider>
```

#### File upload example
```js
// Ensure you pass publicKey, urlEndpoint, and authenticator function to the parent ImageKitProvider component or to the IKUpload component directly.

> <ImageKitProvider

Reviewer 2: Does this SDK not support `checks`? If yes, maybe add it like we have in react etc. Search react docs once and see where have we put checks and make sure we have it all.

Reviewer: yes it supports checks

Reviewer: Done, updated readme.md

  publicKey="your_public_api_key"
  urlEndpoint="https://ik.imagekit.io/your_imagekit_id"
  authenticator={()=>Promise} 
  // This promise  resolves with an object containing the necessary security parameters i.e `signature`, `token`, and `expire`.
  >
  <IKUpload
    onError={onError}
    onSuccess={onSuccess}
  />

  // Passing different upload API options
  <IKUpload
    fileName="file-name.jpg"
    tags={["sample-tag1", "sample-tag2"]}
    customCoordinates={"10,10,10,10"}
    isPrivateFile={false}
    useUniqueFileName={true}
    responseFields={["tags"]}
    folder={"/sample-folder"}
    ref={uploadRef}
    onError={onError} onSuccess={onSuccess}
    transformation={{
      pre: 'l-text,i-Imagekit,fs-50,l-end', 
      post: [
        {
          'type': 'transformation', 
          'value': 'w-100'
        }
      ]
    }}
  />
</ImageKitProvider>
```

## Demo application
* The official step-by-step Next.js quick start guide - https://imagekit.io/docs/integration/next

## Components

The library includes 6 Components:

> * [`ImageKitProvider`](#ImageKitProvider) is used to define options such as `urlEndpoint`, `publicKey`, or `authenticator` for all child components. It does not render any UI elements.

Reviewer 2: Why do we have two components which sort of does same thing i.e. `ImageKitProvider` and `ImageKitContext`

Author: provider and context are two different entities, this change has been done to follow proper react naming conventions and follow the context provider pattern

Author: The provider is to define the values, the context is to consume the values defined values.

* `IKImage` for [image resizing](#image-resizing). This utilizes next/image and renders an `<img>` tag.
* `IKVideo` for [video resizing](#video-resizing). This renders a `<video>` tag.
* `IKUpload`for client-side [file uploading](#file-upload). This renders a `<input type="file">` tag.

> * `ImageKitClient` for [Core SDK](#imagekitclient), This exposes methods from [ImageKit javascript SDK](https://github.com/imagekit-developer/imagekit-javascript) like url and upload.

Reviewer 2: T chota kar lo this ka.

Reviewer: Done

## ImageKitProvider

To use this SDK, you need to provide it with a few configuration parameters. You can use a parent `ImageKitProvider` component to define common options for all children `IKImage`, `IKVideo` or `IKupload` components. For example:

```js
<ImageKitProvider
  urlEndpoint="https://ik.imagekit.io/your_imagekit_id"  // Required. Default URL-endpoint is https://ik.imagekit.io/your_imagekit_id
  publicKey="your_public_api_key" // optional
  transformationPosition="path" // optional
  authenticator={()=>Promise} // optional
  <IKImage path="/default-image.jpg" width={200} height={200} alt="Alt text"/>
</ImageKitProvider>
```

* `urlEndpoint` is required to use the SDK. You can get URL-endpoint from your ImageKit dashboard - https://imagekit.io/dashboard/url-endpoints.
* `publicKey` and `authenticator` parameters are required if you want to use the SDK for client-side file upload. You can get these parameters from the developer section in your ImageKit dashboard - https://imagekit.io/dashboard/developer/api-keys.
* `transformationPosition` is optional. The default value for this parameter is `path`. Acceptable values are `path` & `query`

> > Note: Do not include your [private key](https://docs.imagekit.io/api-reference/api-introduction/api-keys#private-key) in any client-side code.

Reviewer 2: Please replace all docs.imagekit.io links with valid new imagekit.io/docs link.

Reviewer: Done updated the doc links

| transformationPosition | String           | Optional. The default value is `path`, which places the transformation string as a path parameter in the URL. It can also be `query`, which adds the transformation string as the query parameter tr in the URL. If you use the src parameter to create the URL, then the transformation string is always added as a query parameter.                                                                                                                                             |
| queryParameters        | Object           | Optional. These are the other query parameters that you want to add to the final URL. These can be any query parameters and are not necessarily related to ImageKit. Especially useful if you want to add some versioning parameters to your URLs.                                                                                                                                                                                                                                |

> ### Basic video resizing examples

Reviewer 2: Add a paragraph or heading + para about adaptive bitrate streaming and provide this link of the blog - https://imagekit.io/blog/nextjs-video-player/#adaptive-bitrate-streaming

Reviewer: Done, added  a para for ABS in readme.
